sdks/java/core/src/test/java/org/apache/beam/sdk/io/CompressedSourceTest.java

```diff
@@ -632,6 +646,27 @@ protected boolean readNextRecord() throws IOException {
       protected long getCurrentOffset() {
         return offset;
       }
+
+      @Override
+      public Instant getCurrentTimestamp() throws NoSuchElementException {
+        Instant timestamp = ((ByteSource) getCurrentSource()).timestamp;
+        return timestamp == null ? super.getCurrentTimestamp() : timestamp;
+      }
+    }
+  }
+
+  private static class FilterOnTimestamp extends DoFn<Byte, Byte> {
```

Comment:
How about `DoFn<Byte, KV<Long, Byte>>` where we output a KV of the element with its timestamp.

Then we can modify the `PAssert` call in the test with to check bytes with indices/timestamps and ensure that every element gets its correct, distinct timestamp.